File: PerfTester-Xcode/PerfTester-Xcode/Models/MultithreadingTest.swift
//
//  MultithreadingTest.swift
//  PerfTester-Xcode - REL
//
//  Created by Patryk Średziński on 12/11/2018.
//  Copyright © 2018 private.perftesterxcode. All rights reserved.
//

import UIKit

class MultithreadingTest: Test {
    
    override init() {
        super.init()
        title = "Multithreading Test"
        desc = "Creating many threads and waiting for them all to finish"
        imageName = "Multithreading"
        parameters = [ 100, 200, 500, 1000, 2000, 5000, 10000, 20000, 50000, 100000, 200000, 500000, 1000000]
    }
    
    override func prepare(param: Double) {
    }
    
    override func done(param: Double) {
    }
    
    override func doJob(param: Double) {
        let dG = DispatchGroup()
        for _ in 1...Int(param) {
            dG.enter()
            DispatchQueue.main.async {
                dG.leave()
            }
        }
        dG.notify(queue: .main) {
            super.finishJob(param: param)
        }
    }
}

File: PerfTester-Xamarin/PerfTester-Xamarin/Models/MultithreadingTest.cs
using System;
using CoreFoundation;

namespace PerfTesterXamarin.Models
{
    public class MultithreadingTest: Test
    {
        public override string Title => "Multithreading Test";
        public override string Desc => "Creating many threads and waiting for them all to finish";
        public override string ImageName => "Multithreading";
        public override double[] Parameters => new double[] { 100, 200, 500, 1000, 2000, 5000, 10000, 20000, 50000, 100000, 200000, 500000, 1000000 };

        public override void Prepare(double param)
        { }

        public override void Done(double param)
        { }

        public override void DoJob(double param)
        {
            DispatchGroup dG = DispatchGroup.Create();
            for (long i = 0; i < param; i++)
            {
                dG.Enter();
                DispatchQueue.MainQueue.DispatchAsync(() => 
                {
                    dG.Leave();
                });
            }
            dG.Notify(DispatchQueue.MainQueue, () =>
            {
                base.FinishJob(param);
            });
        }

    }
}
